src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -336,11 +359,66 @@ public void CopyChangedValues(TStructuralType original)
                     bool isDeltaType = TypedDelta.IsDeltaOfT(deltaNestedResource.GetType());
                     Contract.Assert(isDeltaType, nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
 
+                    Type nestedType = deltaNestedResource.StructuredType;
+                    Type originalType = originalNestedResource.GetType();
+
+                    if (deltaNestedResource.IsComplexType && !nestedType.IsAssignableFrom(originalType))
+                    {
+                        originalNestedResource = ReAssignComplexDerivedType(original, nestedResourceName, originalNestedResource, nestedType, originalType);
+                    }
+
                     deltaNestedResource.CopyChangedValues(originalNestedResource);
                 }
             }
         }
 
+        private dynamic ReAssignComplexDerivedType(TStructuralType original, string nestedResourceName, dynamic originalNestedResource, Type nestedType, Type originalType)
+        {
+            //As per OASIS discussion, changing a complex type from 1 derived type to another is allowed if both derived type have a common ancestor and the property
+            //is declared in terms of a common ancestor. The logic below checks for a common ancestor. Create a new object of the derived type in delta request.
+            //And copy the common properties.
+
+            Type complexBaseType = nestedType.BaseType;
```

Comment:
>nestedType.BaseType [](http://example.com/codeflow?start=35&length=19)

Please confirm your start states here and on line 390. It looks like you only check for subtypes, not if one of the actual types is a subtype of the other.  What if nestedType derives from originalType or vice-versa?  To validate, please add a test case for changing engine from V4 to V41 and from V41 to V4.

Reply:
Updated. Parent to child already taken care. Added tests too

Reply:
Not convinced your tests cover the scenario I was looking for.

Both the original and new tests have the type of the original value the same as the expected type of the property (I would still like to understand how changing the type works in this case, though, if it doesn't go through your new code).  The new test I was looking for was a property of declared type "V2", whose current value was type V4 and new value V41, and vice-versa.